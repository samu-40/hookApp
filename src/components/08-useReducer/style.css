*{
    margin: 0;
    padding: 0;
    box-sizing: border-box;
}

body{
    padding: 0 50px;
}

.list-group-item > p{
    margin: 0;
    cursor: pointer;
}

.list-group-item{
    display: flex;
    justify-content: space-between;
    align-items: center;
}

.complete{
    text-decoration: line-through;
    color: #555;
}