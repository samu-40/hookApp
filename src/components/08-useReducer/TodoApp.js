import React, { useEffect, useReducer } from 'react';
import { todoReducer } from './todoReducer';
import { TodoList } from './components/TodoList';
import TodoFormAdd from './components/TodoFormAdd';

import './style.css';

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);


    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const handleAdd = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo,
        });
    }

    const handleDelete = (todoId) => {
        const action = {
            type: 'delete',
            payload: todoId,
        };

        dispatch(action);
    };

    const handleToggleDone = (todoId) => {
        dispatch({
            type: 'toggle',
            payload: todoId,
        }); 
    };

    return (
        <div>
            <h1>TodoApp</h1>
            <hr/>

            <h5>TODOS: {todos.length}</h5>
            
            <div className='row'>
                <div className='col-7'>
                    <TodoList 
                        todos={todos}
                        handleDelete={handleDelete}
                        handleToggleDone={handleToggleDone}
                    />
                </div>
                <div className='col-5'>
                    <TodoFormAdd handleAdd={handleAdd}/>
                </div>
            </div>
        </div>
    );

};