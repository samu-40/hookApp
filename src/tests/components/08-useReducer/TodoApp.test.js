import { act } from '@testing-library/react';
import { shallow, mount } from 'enzyme';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Pruebas en TodoApp.js', () => {

    const wrapper = shallow(<TodoApp/>)
    
    Storage.prototype.setItem = jest.fn();

    test('Mostrar el componente correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    
    test('Debe agregar un TODO', () => {
        const wrapper = mount(<TodoApp/>);

        act(() => {
            wrapper.find('TodoFormAdd').prop('handleAdd')(demoTodos[0]);
            wrapper.find('TodoFormAdd').prop('handleAdd')(demoTodos[1]);
        });

        expect(wrapper.find('h5').text().trim()).toBe('TODOS: 2');
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });

    test('Debe eliminar un TODO', () => {
        wrapper.find('TodoFormAdd').prop('handleAdd')(demoTodos[0]);
        wrapper.find('TodoList').prop('handleDelete')(demoTodos[0].id);

        expect(wrapper.find('h5').text().trim()).toBe('TODOS: 0');
    });
    
});
