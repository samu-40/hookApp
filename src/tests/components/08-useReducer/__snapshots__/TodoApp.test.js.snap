// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoApp.js Mostrar el componente correctamente 1`] = `
<div>
  <h1>
    TodoApp
  </h1>
  <hr />
  <h5>
    TODOS: 
    0
  </h5>
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggleDone={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoFormAdd
        handleAdd={[Function]}
      />
    </div>
  </div>
</div>
`;
